src/fparser/two/parser.py

```diff
@@ -197,91 +180,103 @@ def _setup(self, input_classes):
+
+            :returns: names of 'nearest' subclasses with `match` methods.
+            :rtype: List[str | NoneType]
+
+            """
+            if clsname not in base_classes:
+                error_string = f"Not implemented: {clsname}"
+                logging.getLogger(__name__).debug(error_string)
+                return []
+            # remove this code when all classes are implemented.
+            cls = base_classes[clsname]
+            if hasattr(cls, "match"):
+                # This class has a `match` method so no need to search further
+                # down the tree.
+                return [clsname]
+            # clsname doesn't have a `match` method so we look at each of its
+            # subclasses and find the nearest class in each that does have a
+            # `match` method.
+            bits = []
+            for names in getattr(cls, "subclass_names", []):
+                list1 = _rpl_list(names)
+                for names1 in list1:
+                    if names1 not in bits:
+                        bits.append(names1)
+            return bits
+
+        # Dict in which to store optimised list of subclass names for each cls.
+        local_subclass_names = {}
+
+        for cls in base_classes.values():
+            if not hasattr(cls, "subclass_names"):
+                continue
+            # The optimised list of subclass names will only include subclasses
+            # that have `match` methods.
+            opt_subclass_names = []
+            for names in cls.subclass_names:
+                for names1 in _rpl_list(names):
+                    if names1 not in opt_subclass_names:
+                        opt_subclass_names.append(names1)
+            local_subclass_names[cls] = opt_subclass_names[:]
+
+        # Now that we've optimised the list of subclass names for each class,
+        # use this information to initialise the Base.subclasses dictionary:
+        for clsname, cls in base_classes.items():
+            if not hasattr(cls, "subclass_names"):
                 message = "%s class is missing subclass_names list" % (clsname)
                 logging.getLogger(__name__).debug(message)
                 continue
+            subclass_names = local_subclass_names.get(cls, [])
             try:
-                bits = fparser.two.Fortran2003.Base.subclasses[clsname]
+                bits = Fortran2003.Base.subclasses[clsname]
             except KeyError:
-                fparser.two.Fortran2003.Base.subclasses[clsname] = bits = []
+                Fortran2003.Base.subclasses[clsname] = bits = []
             for name in subclass_names:
                 if name in base_classes:
                     bits.append(base_classes[name])
                 else:
-                    message = "{0} not implemented needed by {1}".format(name, clsname)
+                    message = f"{name} not implemented needed by {clsname}"
                     logging.getLogger(__name__).debug(message)
 
         if 1:
```

Comment:
You removed the previous "if 1:". Can you also remove this one and the commented code inside (unless is something we need to keep) and convert to f-string.

Reply:
Oh yes, tidied. I've not converted to f-string as I believe the % formatting is the preferred thing to do with logging?

Reply:
Actually, I realise in this case the formatting was already happening so I have changed to f string.